Make application version and name available as (fixed) runtime properties

geoand opened the pull request:
Fixes: #6255

Relates to: #6267

geoand committed:
Make application version and name available as runtime properties

Fixes: #6255

Co-authored-by: Guillaume Smet <guillaume.smet@gmail.com>

geoand commented:
I am making this a draft because it only solves part of the problem. With this change `quarkus.application.name` and `quarkus.application.version` are available when running the prod mode of the application. It doesn't work for dev-mode or tests.
@dmlloyd I can dig further, but maybe you have some insight why this is?

gsmet commented:
We need to add a test for that one.

geoand commented:
Yes for sure. However at the moment it only works for prod mode so the only test we can have is one of the heavyweight tests that invoke Maven.

geoand commented:
Never mind, I think I have a fix for dev mode as well.

geoand commented:
Actually this needs more thought since with my fix you can't override the value in dev-mode.

geoand commented:
@gsmet had the idea to use the build system properties introduced in https://github.com/quarkusio/quarkus/pull/3137.

This is well worth exploring so I'll do that in a few hours and see what happens

geoand committed:
Add tests for application name and version handling

gsmet commented on core/runtime/src/main/java/io/quarkus/runtime/ApplicationConfig.java:
```diff
@@ -0,0 +1,31 @@
+    public String version;
+
+    public static boolean isNameUnset(String name) {
+        return name.contains("unset");
```
shouldn't it be `equals()` instead of contains? What if you have `unset` in your project name or even if the artifact is called `unset` (which might be a nice project name)?

I would make it `__UNSET__` or something similar.

geoand replied:
Yes you are right. This got left over from previous stuff I was checking. Let me fix it

geoand replied:
Done

gsmet reviewed (changes requested):
I added a small suggestion. It's nice to have tests for that one.

I think what we need now is a Gradle equivalent.

geoand commented:
prod mode of Gradle will work, so if we get this in I am hoping @gastaldi will take care of the necessary changes for dev mode :)

geoand committed:
Ensure that NativeIT is only run during when using native profile

The other tests are only run when this profile is not active in order
to avoid running them multiple times (since they are slow as well)

Co-authored-by: Guillaume Smet <guillaume.smet@gmail.com>

geoand commented:
I'm taking this out of draft since I also fixed some CI issues the PR had

dmlloyd commented on core/runtime/src/main/java/io/quarkus/runtime/ApplicationConfig.java:
```diff
@@ -0,0 +1,31 @@
+     * If not set, defaults to the name of the project.
+     */
+    @ConfigItem(defaultValue = "${project.name:__UNSET__}")
+    public String name;
```
If we allow it to be unset, it should just be optional.

geoand replied:
Can a property have a default value and also be `Optional`?

geoand replied:
The reason I went this was is because of https://github.com/quarkusio/quarkus/blob/master/core/deployment/src/main/java/io/quarkus/deployment/configuration/RunTimeConfigurationGenerator.java#L738. 
With the current code of `ApplicationConfig` in master the line above is what (I believe) causes the value to not be used at runtime.

dmlloyd replied:
> Can a property have a default value and also be `Optional`?

Yes, it can.  But the default value should be meaningful.  It looks like you might be trying to hack around something the wrong way here.

> The reason I went with this is, is because of https://github.com/quarkusio/quarkus/blob/master/core/deployment/src/main/java/io/quarkus/deployment/configuration/RunTimeConfigurationGenerator.java#L738.
> With the current code of `ApplicationConfig` in master the line above is what (I believe) causes the value to not be used at runtime.

This logic essentially means "if this property has no default value, and no other sub-trees of this value have default values, don't generate pointless default value methods which would always return null for every name".  This logic only controls what values end up as defaults at run time based on what's in the parse tree.  If a property has no default value, then no default value needs to be returned for it.

This logic does *not* generate run-time-readable config values for values which were specified in build time config, which I suspect is what you're trying to do here.  Trying to use this logic for that purpose will yield incorrect results.

geoand replied:
Well I wasn't trying to use something in any particular way. I was lead to that piece of code by debugging and comparing what made this type of configuration different from others that did behave the way we expected. 
But let me go back and take another look.

geoand replied:
> This logic does _not_ generate run-time-readable config values for values which were specified in build time config, which I suspect is what you're trying to do here. 

And yes, that is exactly the end result I am looking for :)

geoand replied:
Thanks for the invaluable input! I think this iteration is better. WDYT?

gastaldi commented on core/runtime/src/main/java/io/quarkus/runtime/ApplicationConfig.java:
```diff
@@ -0,0 +1,31 @@
+    public String version;
+
+    public static boolean isNameUnset(String name) {
+        return name.equals("__UNSET__");
```
```suggestion
        return (name == null) || ("__UNSET__".equals(name));
```

gastaldi replied:
For the cases where the Config object is uninitialized properly

geoand replied:
I'm going to change this stuff based on @dmlloyd's comments, but thanks for the suggestion

geoand commented:
@dmlloyd what do you think about the latest iteration?

dmlloyd reviewed (approved):
Seems like a legitimate incremental improvement.

geoand commented:
Thanks!

geoand commented:
@gsmet OK to merge?

gsmet commented on devtools/maven/src/main/java/io/quarkus/maven/DevMojo.java:
```diff
@@ -490,7 +490,13 @@ void prepare() throws Exception {
+
             devModeContext.getBuildSystemProperties().putAll((Map) project.getProperties());
+
+            // this is a minor hack to allows up to populate ApplicationConfig with default
```
```suggestion
            // this is a minor hack to allow ApplicationConfig to be populated with defaults
```

geoand replied:
Fixed, thanks

gsmet commented on integration-tests/maven/src/test/java/io/quarkus/maven/it/NativeImageIT.java:
```diff
@@ -16,13 +16,13 @@
- *
- */
+// ensure that this test only runs when GraalVM is configured
+@EnabledIfEnvironmentVariable(named = "GRAALVM_HOME", matches = ".*")
```
So we don't run it on CI with Docker? I don't think GRAALVM_HOME is set in this case?

geoand replied:
We do, I checked the docker image that we use and it does set it

gsmet replied:
The Docker image sets it. But is it set when the test is executed?

geoand replied:
It's part of the Docker image, so it's present when the container is run

gsmet replied:
I might be missing something but my understanding is the test is not executed inside the Docker image and thus the `GRAALVM_HOME` is not set. Am I missing something?

geoand replied:
Aha, I see what you are saying! Indeed you are absolutely right! I'll try and think of some other way to do this.
However I don't knoq how this test ever passed in the first place... The reason I disabled it is because the PR was failing on it, because of course native-image doesn't exist

geoand replied:
I pushed a change. WDYT?

geoand replied:
Fixed a different way

gsmet reviewed (changes requested):
I added a suggestion and a question.

gsmet commented:
BTW, you got back to a previous version? I don't see the thing we talked about.

geoand commented:
> BTW, you got back to a previous version? I don't see the thing we talked about.

I changed it based on the feedback from @dmlloyd. It essentially does the use the build parameters, but for setting the application name and version directly.

gsmet commented on integration-tests/maven/src/test/java/io/quarkus/maven/it/NativeImageIT.java:
```diff
@@ -37,8 +34,13 @@ public void testJavaLibraryPathAtRuntime() throws Exception {
+        if (System.getenv("GRAALVM_HOME") == null || System.getenv("GRAALVM_HOME").isEmpty()) {
+            // If the GRAALVM_HOME env var isn't set, try to use docker to build the image.
+            // This is needed because CI doesn't have a GraalVM installation
+            mavenArgs.add("-Dquarkus.native.container-runtime=docker");
```
Can't say I'm excited about this because that means we will require Docker to be installed for a standard JDK run? Or am I mistaken?

geoand replied:
Well unless I have completely missread this test, it can't succeed without GraalVM being installed, right?
So the change just allows us to fallback to docker if it's not.

geoand replied:
So essentially now it's safer than it was before, right?

geoand replied:
Fixed a different way

geoand commented:
It seems like there is an issue with the Windows test.... Seems like Docker on Windows isn't playing nicely in CI...
Should we perhaps ignore this test for Windows?

geoand commented:
> Should we perhaps ignore this test for Windows?

This is what I ended up doing

gastaldi commented on devtools/maven/src/main/java/io/quarkus/maven/DevMojo.java:
```diff
@@ -490,7 +490,13 @@ void prepare() throws Exception {
             devModeContext.getBuildSystemProperties().putAll((Map) project.getProperties());
+
+            //  this is a minor hack to allow ApplicationConfig to be populated with defaults
+            devModeContext.getBuildSystemProperties().putIfAbsent("quarkus.application.name", project.getArtifactId());
```
Should it resolve to `project.name` first and fallback to `artifactId`?

geoand replied:
Does the former exist for Gradle as well?

Also note that we already do the same thing for the regular build, it was just missing from dev mode

gastaldi replied:
> Does the former exist for Gradle as well?

I have no idea, but now that you mention the `QuarkusDev` task in the Gradle plugin needs to be updated too, no?

> Also note that we already do the same thing for the regular build, it was just missing from dev mode

:+1:

geoand replied:
It does yes, I was hoping you would take that up once this is merged 😉

geoand commented:
@gsmet unless you have any suggestions for getting around the NativeImageIT problem some other way, should we merge this?

geoand commented:
PR updated with a suggestion from @gsmet regarding better handling of the problematic `NativeImageIT`

gsmet reviewed (approved):
As mentioned on Zulip, I'm not entirely convinced by the annotation approach (especially the fact that it was needed) but since it's there, let's make progress on this one.
